gnovm/pkg/gnolang/machine.go

```diff
@@ -2024,117 +2024,105 @@ func (m *Machine) Printf(format string, args ...interface{}) {
+		builder.WriteString(fmt.Sprintf("          #%d %v\n", i, m.Exprs[i]))
 	}
-	bs := []string{}
+
+	for i := len(m.Stmts) - 1; i >= 0; i-- {
+		builder.WriteString(fmt.Sprintf("          #%d %v\n", i, m.Stmts[i]))
+	}
+
 	for b := m.LastBlock(); b != nil; {
-		gen := len(bs)/3 + 1
+		gen := builder.Len()/3 + 1
 		gens := "@" // strings.Repeat("@", gen)
+
 		if pv, ok := b.Source.(*PackageNode); ok {
 			// package blocks have too much, so just
 			// print the pkgpath.
-			bs = append(bs, fmt.Sprintf("          %s(%d) %s", gens, gen, pv.PkgPath))
+			builder.WriteString(fmt.Sprintf("          %s(%d) %s\n", gens, gen, pv.PkgPath))
 		} else {
 			bsi := b.StringIndented("            ")
-			bs = append(bs, fmt.Sprintf("          %s(%d) %s", gens, gen, bsi))
+			builder.WriteString(fmt.Sprintf("          %s(%d) %s\n", gens, gen, bsi))
+
 			if b.Source != nil {
 				sb := b.GetSource(m.Store).GetStaticBlock().GetBlock()
-				bs = append(bs, fmt.Sprintf(" (s vals) %s(%d) %s", gens, gen,
-					sb.StringIndented("            ")))
+				builder.WriteString(fmt.Sprintf(" (s vals) %s(%d) %s\n", gens, gen, sb.StringIndented("            ")))
+
 				sts := b.GetSource(m.Store).GetStaticBlock().Types
-				bs = append(bs, fmt.Sprintf(" (s typs) %s(%d) %s", gens, gen,
-					sts))
+				builder.WriteString(fmt.Sprintf(" (s typs) %s(%d) %s\n", gens, gen, sts))
 			}
 		}
-		// b = b.Parent.(*Block|RefValue)
+
+		// Update b
 		switch bp := b.Parent.(type) {
 		case nil:
 			b = nil
-			break
 		case *Block:
 			b = bp
 		case RefValue:
-			bs = append(bs, fmt.Sprintf("            (block ref %v)", bp.ObjectID))
+			builder.WriteString(fmt.Sprintf("            (block ref %v)\n", bp.ObjectID))
 			b = nil
-			break
 		default:
 			panic("should not happen")
 		}
 	}
-	obs := []string{}
+
 	for i := len(m.Blocks) - 2; i >= 0; i-- {
 		b := m.Blocks[i]
+
+		if b == nil || b.Source == nil {
+			continue
+		}
+
 		if _, ok := b.Source.(*PackageNode); ok {
 			break // done, skip *PackageNode.
 		} else {
-			obs = append(obs, fmt.Sprintf("          #%d %s", i,
+			builder.WriteString(fmt.Sprintf("          #%d %s", i,
 				b.StringIndented("            ")))
 			if b.Source != nil {
 				sb := b.GetSource(m.Store).GetStaticBlock().GetBlock()
-				obs = append(obs, fmt.Sprintf(" (static) #%d %s", i,
+				builder.WriteString(fmt.Sprintf(" (static) #%d %s", i,
 					sb.StringIndented("            ")))
 			}
 		}
 	}
-	fs := []string{}
+
 	for i := len(m.Frames) - 1; i >= 0; i-- {
-		fr := m.Frames[i]
-		fs = append(fs, fmt.Sprintf("          #%d %s", i, fr.String()))
+		builder.WriteString(fmt.Sprintf("          #%d %s\n", i, m.Frames[i]))
 	}
-	rlmpath := ""
+
 	if m.Realm != nil {
-		rlmpath = m.Realm.Path
-	}
-	exceptions := make([]string, len(m.Exceptions))
-	for i, ex := range m.Exceptions {
-		exceptions[i] = ex.Sprint(m)
-	}
-	return fmt.Sprintf(`Machine:
-    CheckTypes: %v
-	Op: %v
-	Values: (len: %d)
-%s
-	Exprs:
```

Comment:
Where does the new code print these headers? I can see one for Realm for not for the others